### PR TITLE
Add “scroll to bottom button”, improve scrolling behavior.

### PR DESCRIPTION
**Pull request by charlesmchen**
…ove scrolling behavior.

PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
https://github.com/WhisperSystems/Signal-iOS/pull/2166/files?w=1

**Comment by charlesmchen** [comment]
@michaelkirk Sorry clang-format had a field day after I extracted the non-move changes out of https://github.com/WhisperSystems/Signal-iOS/pull/2167

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 1]
```diff
@@ -2977,10 +3046,85 @@ - (void)clearUnreadMessagesIndicator
...
-        [self ensureThreadOffersAndIndicators];
+        [self ensureDynamicInteractions];
+    }
+}
+
+- (void)viewDidLayoutSubviews
+{
+    [super viewDidLayoutSubviews];
+
+    [self updateLastVisibleTimestamp];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: This is to ensure that the "scroll to bottom" indicator is moved & updated appropriately when the input toolbar appears, disappears or is resized, e.g. when user input wraps to a new line and the input toolbar grows to accommodate it.

Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: `updateLastVisibleTimestamp` isn't free, but not it's expensive either.

Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: Cool. If it becomes a perf issue we might want to look into:

```
- (void)keyboardController:(JSQMessagesKeyboardController *)keyboardController keyboardDidChangeFrame:(CGRect)keyboardFrame;
```

**Pull request merged**